jsonschema-core.xml

```diff
@@ -455,6 +455,68 @@
+            </t>
+            <!-- Links will be added to relevant spec parts after the sections are reorganized
+                 to create the new schema loading section -->
+            <t>
+                <list>
+                    <t>
+                        All syntactically legal JSON documents with well-defined behavior are fully
+                        supported as instance data.  Notably, JSON objects with duplicate property
+                        names are syntactically legal but not well-defined.
+                    </t>
+                    <t>
+                        JSON Schema is defined over a data model, not JSON text.  Therefore
+                        JSON Schema can be used with any data format that can be reliably parsed
+                        into the data model.
+                    </t>
+                    <t>
+                        Locations in schemas and instances are always identified by URIs,
+                        JSON Pointers, or Relative JSON Pointers.  JSON Schema treats URIs as
+                        identifiers only.  Locating and retrieving URI-identified resources
+                        is outside of the scope of this specification.
+                    </t>
+                    <t>
+                        Schema URIs referenced or defined in JSON Schema's Core vocabulary MUST be
+                        resolvable without instance data.
+                    </t>
+                    <t>
+                        Once schema URIs are resolved, the result of applying a schema object
+                        to instance data is data is a function of that schema object and its
+                        dynamic-scope subschemas.  Parent schemas have no effect on results.
+                    </t>
+                    <t>
+                        Each keyword's behavior falls into one or more behavioral classifications.
+                        A keyword's classification MUST be the same for all values of the keyword.
+                    </t>
+                    <t>
+                        Keywords produce a boolean assertion result, and an annotation result that
+                        can be of any type.  Depending on the classification, the assertion result
+                        might always be true, or the annotation result might always be empty.
+                    </t>
+                    <t>
+                        Keyword results MAY be defined in terms of the results of other keywords
+                        in the same schema object, as long as no cyclic dependencies are produced.
+                    </t>
+                    <t>
+                        JSON Schema assertions form a constraint system.  The empty schema,
+                        <spanx style="verb">{}</spanx>, allows everything and constrains nothing.
+                        Each assertion keyword adds constraints; keywords MUST NOT remove constraints.
```

Comment:
"MUST NOT remove *or alter* contraints."

Also, I still think that we should mention that this does not conform to inheritance models such as polymorphism.

Reply:
What sort of "alter" do you have in mind?  99% sure I agree, just want to be clear if something has come up in practice.

And yeah, probably worth a comment.

Reply:
I think that's me drifting off into an inheritance mindset where you can override members.

The case I can think of is redefining something like `maximum` alongside a `$ref` to something that has `maximum` defined within.  But in that case, the lower one would win out because the instance would just have to satisfy both `maximum`s.  Similar for property definitions: the instance would have to validate against both definitions.

This is fine, but it doesn't do what people thinking about inheritance would think it should do.

Reply:
Hmm.... yeah, that's an important point, I'll see if I can come up with concise wording.

Reply:
The trouble is, we DO have keywords which alter the behaviour of other keywords, so I don't think you can say that.

Reply:
@Relequestual There's a wording that will work in here somewhere, but that's part of why I questioned "alter" as a word choice.  Just to make sure we're all on the same page about the issue, I think the simplest version of this is:

```JSON
{
    "type": "object",
    "additionalProperties": {"type": "integer"}
}
```
vs
```JSON
{
    "type": "object",
    "properties": {"foo": true},
    "additionalProperties": {"type": "integer"}
}
```
where the `properties` keyword technically adds a constraint that is specific to `foo`, but because that carves `foo` out of `additionalProperties`, from the point of view of that instance property the constraints are _changed_, and given that the schema used for `foo` is `true`, the effect is to remove the `"type": "integer"` constraint.

Reply:
There might be something here about applicators vs assertions.  You can use applicators to change which subschema(s) _apply_ to a location in the instance.  But adding assertions to a given schema object always further constrains.  The only way to effectively lift a constraint is to use applicators to change which schema object is used.

Reply:
> but because that carves `foo` out of `additionalProperties`, from the point of view of that instance property the constraints are *changed*...

I don't consider this as `additionalProperties` changing its behavior.  `additionalProperties` is *defined* to depend on the existence of `properties`, so the behavior of seeing `foo` only when it's not handled by `properties` is consistent with that definition.

Maybe "override" would be a better word here?

Reply:
I intentionally didn't say (or at least didn't mean to say) that `additionalProperties` changed its behavior.  Rather, from the black box validation point of view, a constraint appeared to be lifted when a keyword was added.  In a proper constraint layering system, the set of valid instances after adding any keyword would be a subset of the set of valid instances before adding the keyword.

However, this instance: `{"foo": "hello", "bar": 2}` fails validation against my first example schema, because `"hello"` is not an integer.  But it _passes_ validation against my second example schema, because _adding_ `properties` caused a different subschema (`true`) to apply to `"hello"`.

I'm 99% certain that if we just look at assertions, adding an assertion always reduces the set of valid instances (ignoring complex situations like `"not"` where adding to its subschema reduces that subschema's valid set, which expands the inverted set accepted by `"not"`.  Hmm... that's an applicator behavior, too.  Not the "which subschema applies to what" part but the "perform boolean logic on the subschema results to produce the applicator result."

I think it is always true that adding a pure **assertion** keyword (that is not also an applicator and/or annotation) to a schema object always reduces the set of valid instances according to that schema object (including its subschemas, because subschemas are never affected by parent schemas once you resolve URI references- that's one of the other principles).

Applicators allow much more complex behavior, but it's still just changing how the subschemas **results** are combined.  This is why there is never any override or merge or replacement.  The operations are performed on subschema results, never on individual keywords.  Aside from information returned in annotation results, subschemas are black boxes; the parent schema has no idea what keywords they contain.

Reply:
Could the issue here be it's "assertions and annotations must not remove constraints" but applicator keywords will alter how constraints are applied?

`properties` and `additionalProperties` are both applicators.
An argument could be made that when using `if/then/else`, the instance data actually modifies how the constraints are applied also. BUT, they are not removed, just applied differently.

Maybe it's worth noting that while constraints cannot be removed, the application of the schema they belong to may be altered as a side effect of the annotation results of other keywords?

Reply:
@Relequestual yeah what you're saying here is getting very close to it, I think.